### ISSUE TITLE
Consider adding Meson build system support

### ISSUE THREAD

**Issue by nazar-pc** [openssl/openssl#16812]
This is a spin-off from #10902 that is CMake-specific, at least it seems to be, while I decided to start an official separate discussion for Meson as well. The discussion in #10902 is relevant if someone is brave enough to read it, but basically OpenSSL is challenging to use in many cases because of:
* obscure build system
* reliance on Perl

Alternative and modern build system could be very beneficial and Meson is one such modern option (and it is Python-based, which in turn can be used to eventually replace Perl for scripting purposes).

This was initially posted as https://github.com/openssl/openssl/issues/10902#issuecomment-940781514, relevant discussions include #10752 and #16188.

### Background
I am not affiliated with any of the build systems, OS environments, etc. But I am the author of the Meson OpenSSL wrap as it is right now because I needed it for a project, so I can share a bit of feedback about that.

There is [a project](https://github.com/versatica/mediasoup) where GYP build system was used and there was a ticket open for a few years to migrate to CMake. I decided to tackle it and it was a very frustrating experience until I gave up and tried Meson: https://twitter.com/nazarpc/status/1413682286118285316

From there I was able to not only get things working in a reasonable amount of time with Meson, but eventually upstream most of my work on the project that resulted, among other things, in OpenSSL Meson wrap we have today.

### How to use OpenSSL with Meson *today*
Since ease of use is a critical concern that people have when discussion starts about new build system, I though I'd give you a full end-to-end example of working with Meson and current OpenSSL wrap so you can judge it based on your experience with whatever build system you're using right now.

Step 1: grab meson and ninja, create project folder
```bash
pip3 install meson ninja
mkdir openssl-test
cd openssl-test
```

Step 2: install OpenSSL wrap, we'll use it instead of whatever provided by the system (if any)
```bash
mkdir subprojects
meson wrap install openssl
```

Step 3: Let's have a simple application
```c
// Complete openssl_test.c:
#include <stdio.h>
#include <openssl/crypto.h>

int main(int argc, char **argv) {
    printf("OpenSSL version is %s.\n", OpenSSL_version(OPENSSL_VERSION));
    return 0;
}
```

Step 4: and `meson.build` for our little project
```meson
# Complete meson.build:
project('openssl-test', 'c')

executable('openssl_test', 'openssl_test.c',
  dependencies: [
    subproject('openssl').get_variable('openssl_dep')
  ]
)
```

Step 5: compile and run the project
```bash
meson setup build
meson compile -C build
build/openssl_test
```

Example:
```
nazar-pc@nazar-pc:/f/openssl-test> build/openssl_test
OpenSSL version is OpenSSL 1.1.1l  24 Aug 2021.
```

I **really** wish one day this will be the upstream experience.

**Comment by kroeckx** [comment]
Can you clarify if this is an replacement of our current system,
or an addition to our system that we would additionally need to
maintain?

It seems to me that meson's reliance on python is just as
problematic or even more so than our current reliance on perl.

Note that perl is used for various things currently:
- Generating the Makefiles
- Generating assembler files
- Test suite
- Various other scripts

If you want to remove the need for perl, you will need to either
have a way so that it's not always needed during a build, or have
also have a replacement/alternative implementation.

If the intend is to replace the build system, this will probably
seriously reduce the number of supported platforms.

It's also not clear to me what issue you're really trying to
solve. You're suggesting a solution without clearly stating
what your problem is. The problem might somehow have something
to do with perl, but it's not clear.

**Comment by xclaesse** [comment]
> The requirements to be able to build python are much higher than those to build perl. So perl is available on much more platforms.

[citation needed] :D But it could be true indeed.

> I'm not sure that python is an improvement.

In my opinion it's a huge improvement in readability of the code, but I guess that's a matter of taste and I'm biased :p

**Comment by xclaesse** [comment]
> I'm not sure what python is written in, but ninja seems use C++, which or might not be available on some of those platforms.

IIRC it does not require c++11 unlike cmake, so it could be not that bad. But I could be wrong.

**Comment by diamondman** [comment]
@kroeckx  
> Note that there are plenty of platforms you've never heard about, where OpenSSL is currently used.

Do you know of any off hand that do not support Python3? Curious to see how much effort it would be to add support.

**Comment by levitte** [comment]
> @kroeckx
> 
> > Note that there are plenty of platforms you've never heard about, where OpenSSL is currently used.
> 
> Do you know of any off hand that do not support Python3? Curious to see how much effort it would be to add support.

Python3 exists for VMS, but I haven't checked yet how well the VMS filespec syntax is supported, among a few things that I know can be challenging.
Top of my head, I can guess that NonStop systems are among these as well. @rsbeckerca might have a more accurate answer.

Essentially, I have low expectation on support for anything but the 3 big: Unix family, MacOS and Windows. Investigation needed...

**Comment by rsbeckerca** [comment]
> > @kroeckx
> > > Note that there are plenty of platforms you've never heard about, where OpenSSL is currently used.
> > 
> > 
> > Do you know of any off hand that do not support Python3? Curious to see how much effort it would be to add support.
> 
> Python3 exists for VMS, but I haven't checked yet how well the VMS filespec syntax is supported, among a few things that I know can be challenging. Top of my head, I can guess that NonStop systems are among these as well. @rsbeckerca might have a more accurate answer.
> 
> Essentially, I have low expectation on support for anything but the 3 big: Unix family, MacOS and Windows. Investigation needed...

@kroeckx @levitte Support for Python 3 on NonStop is available on x86 only. The NonStop ia64 variant only has, and will only have, Python 2. NonStop ia64 is still supported for the next few years but will disappear somewhen after the 3.0.x deprecation timeframe. Python 2 is also available on x86 but is considered mature (so I would have to work around any issues, like the `env -i` we found earlier this year). Note that 3.0.x adoption has been very low to date.

**Comment by EvanCarroll** [comment]
@rsbeckerca do you have any evidence that NonStop OS is for ia64 is still supported? re:

> NonStop ia64 is still supported for the next few years but will disappear somewhen after the 3.0.x deprecation timeframe.

**Comment by rsbeckerca** [comment]
> @rsbeckerca do you have any evidence that NonStop OS is for ia64 is still supported? re:
> 
> > NonStop ia64 is still supported for the next few years but will disappear somewhen after the 3.0.x deprecation timeframe.
> 
> I don't see a published support cycle, or EOL date. The nonstop-os website is dead for me, and was last updated in 2016 (per the footer) http://www1.hp.com/us/en/enterprise/servers/integrity/nonstop/nonstop-os.aspx

NonStop is definitely not dead. The link is no longer useful since HP split into HP (Consumer) and HPE (E=Enterprise). The link describing NonStop is at [https://www.hpe.com/us/en/servers/nonstop.html](https://www.hpe.com/us/en/servers/nonstop.html). Current versions of the operating system run at least through late 2027, but that is likely to be pushed out as new versions of the operating system are released (a new version is expected to be released imminently). There is no end of life published for the current x86 platform variants. The most recent ia64 variant support goes to at least the end of 2025 but there is also no end of life date currently published. The operating system ships with a custom build (non-externalized version) of OpenSSL for the platform web server.

1.1.1x, 3.0.x, and 3.1.x (when officially released) builds of OpenSSL for NonStop are done by my team and made available to the NonStop community.

**Comment by EvanCarroll** [comment]
While I'm not trying to impugn your reputation, it's not really actionable information imho unless you can point to an official maintenance schedule. Which shouldn't be too much to ask of HPE, as every free open source distribution does just that:

* Ubuntu has a nice graphic, but you can see more information at https://ubuntu.com/about/release-cycle ![Screenshot_2022-09-20_19-57-18](https://user-images.githubusercontent.com/19922/191390731-6571f86e-38d6-4582-bfe9-0693ab373bcd.png)
* Debian has this page https://wiki.debian.org/DebianReleases

Can we cite the 2027 date somewhere? Also what does it mean to have NonStop ia64 "supported" which uses Python 2, when Python 2 itself is not supported? That is to say, the Python 2 interpreter and standard library have unpatched security vulnerabilities. ActiveState, a company that has a whole service dedicated to Python2 support has a whole team dedicated to backporting security patches for the entire ecosystem the list the bugs they've had to patch here https://www.activestate.com/products/python/python-2-end-of-life-security-updates/ Which IMHO is a pretty abbreviated list, 

To what degree does HPE support Python 2 on NonStop? Is HPE doing this too? I'm just trying to figure out if the right frame here is that they're maintaining Python 2, or if they're just not providing Python 3.

**Comment by nazar-pc** [comment]
I feel like NonStop OS on ia64 or similar discussions are off-topic here. No one suggests replacing existing build system with Meson here, so nothing will change for existing users if/when Meson support is introduced, they can continue using existing build system just like they did before.

**Comment by levitte** [comment]
Fair warning: If discussions are becoming repeatedly disrespectful, it might be time to lock this issue.

/admin

**Comment by t-j-h** [comment]
@rsbeckerca your efforts (and those of all our volunteer community platform maintainers) are very much appreciated - many people (myself included) contribute to OpenSSL for the benefit of the community (and not in a paid-by-our-employer capacity).

**Comment by eli-schwartz** [comment]
AFAICT the only reason it came up at all was because in a project for downloading youtube videos, openssl was brought up as an example of how some projects do actually care about supporting old or exotic platforms, and this is worrying to python programmers who don't like to encourage people to use old software releases as it may drag down the python ecosystem's mindshare.

This is probably not the OpenSSL project's chief concern, and for good reason... the scopes of each project are completely different. OpenSSL is critical infrastructure and it needs to run on more than just what is most time-effective for a one-person developer team to support. It's amazing that people like @rsbeckerca put in that effort. :)

tl;dr it seems like there was just a bit of confusion here.

**Comment by levitte** [comment]
Nicely put, @eli-schwartz

**Comment by EvanCarroll** [comment]
I don't see how disrespectful nor inappropriate to ask for a source pertaining to claim that was made, and to even give examples of what I was looking for. Even going so far as to say I'm **not** trying to "impugn your reputation." I care about supporting exotic and old platforms too. I just want to know what it means when it is said,

* Foo "supports" bar.

And, I think when a specific period is cited "late 2027" to those of us concerned about support, we should be able to find official documentation online. On the contrary, I looked I don't see anything much about "NonStop OS" on the HPE site. There are three mentions of it,

* https://www.google.com/search?q=site%3Ahttps%3A%2F%2Fwww.hpe.com%2Fus%2F+%22NonStop+OS%22

I also can't find any information about what "NonStop OS" supports, how it's licensed, where to get the operating system, or how the package manager works.

So to be absolutely clear, none of this is personal -- I tried to explain that before. I'm just interested in your statement and can't find _any_ information on HPE. You seem like you could point me in the right direction.

**Comment by Tachi107** [comment]
@levitte I'm sorry, I shouldn't have linked this thread from the youtube-dl one. I'll remove the link, but it may be appropriate to lock this thread for a couple of days :/

**Comment by MILASTARX84** [comment]
whats all the hype about  things get updated all the ti e?!!

**Comment by michael-o** [comment]
To me, the OpenSSL team was much more receptible to accept my patches for HP-UX than the Python team. I have up providing patches for Python on HP-UX because they are getting ignored. Perl runs really everywhere.

**Comment by nhorman** [comment]
Marking as inactive, will be closed at the end of 3.4 dev barring further input.

If there is interest in developing a meson/ninja based build environment for openssl, I expect this should likely be a community issue, though given its age, im more inclined to just close it

**Comment by Croydon** [comment]
I am personally more in favour of #10902 than a Meson build system, but nonetheless, I find it a bit weird to close this due to inactivity. What is required here is a clear statement of the OpenSSL team if they would accept such pull requests that add another build system or not and if yes, under what circumstances. Until then, I think that in both issues most things are said so far.

**Comment by nazar-pc** [comment]
> Marking as inactive, will be closed at the end of 3.4 dev barring further input.

What kind of additional input is expected here?

We just want to have modern easy to use build system that integrates well with other tooling. Unfortunately today OpenSSL build system is neither of those things and lack of upstream support for such initiatives even when someone is willing to contribute it is frankly frustrating.

**Comment by levitte** [comment]
So here are some options:

1. Choose *one* build system:

   - there's the current one
   - there's the cmake proposal: #10902 
   - there's the meson proposal (this issue)
   - while I'm at it, I can also throw in [Premake](https://premake.github.io/), a Lua based system that I've been eyeing.

   what to choose...  and how to deal with the platforms that don't have that particular system, or can't (lack of compiler, for example).  There's really no clear winner.

2. Choose multiple systems.  This could be doable, if they were set up in such a way that they all get their information (what files to compile) from the same spot (*the maintainers will **not** be happy if they have to maintain the same information in multiple places!*).  If they'd all read and understand `build.info` files, that could be doable, for example.

3. Perl.  This is not really just tied to the build system.  If you want to get rid of perl dependencies, there's a *lot* of work to do, much much more than just the build system.

These are points that remain unanswered, and that's where all these discussions end up being stuck.

**Comment by levitte** [comment]
Regarding integration with other tooling, that's not the same thing as adopting a particular build system, is it?

There's been some work done to integrate with `pkg-config` since what seems like forever at this point, and lately, with cmake (by producing installable OpenSSLConfig.cmake and OpenSSLConfigVersion.cmake).  We can certainly do more of that sort.

**Comment by nhorman** [comment]
@Croydon @nazar-pc I understand your frustration, but please keep in mind, no open source project (or any project in general) is going to make any sort of promise to accept a PR until we can see that PR, you have to take on a certain amount of faith that your work will get fairly reviewed, and iterate over the feedback until a good consensus is reached (whiich may be that the proposal isn't achievable).  

I understand wanting a more, I suppose for a lack of a better word, standardized build system (I certainly wouldn't complain if we had one), but there are requirements for such a build system that we have to adhere to, which may make any subsequent proposal unworkable.

If you are looking for what those requirements are, I would think a good starting point would be:

1) Single source build deps - as @levitte notes, its imperative that we have a single source of build dependencies which to update on future changes.  Having multiple sources is a maintenance nightmare, and not something we would consider.

2)   Platform Support - If the goal is to replace the existing build system, then the new build system needs to be able to build all the targets for all our platforms successfully (targets listed [here](https://www.openssl.org/policies/general-supplemental/platforms.html)).  The Primary, Secondary and Community platforms are required.  If this effort is meant to produce an adjunct build environment (much like the cmake output we currently have), this could be relaxed somewhat, as the existing build system could be used in an official capacity, while the new build system could be used opportunistically by interested parties.

3) CI integration - As above, if replacement is the goal, then the new build system would need to incorporate changes to the CI workflow files to be used properly, and exercise the buiild system regularly.  Again, as above, if the goal were to produce an opportunistic secondary build system, this could be relaxed.

4) Ongoing maintenence - If your goal is to produce a secondary build system  (i.e. leave the currently build system in place, and create a secondary build system), then I (speaking only for myself), would ask that a community member volunteer to be responsible for its upkeep going forward.  As a secondary build system, its not something we would use regularly, and so the core openssl devs would not be positioned well to maintain it, necessitating a community maintainer.

If you're interested in moving forward under those requirements, I'm happy to move this to a community support issue, which you are welcome to work on (and your contribution is certainly appreciated).  I would suggest, given the size of this effort, that you submit a request for a feature branch, to which you can merge work incrementally without affecting the release cycle, that we could review, and discuss on an ongoing basis

**Comment by nazar-pc** [comment]
@levitte @nhorman I think the questions and concerns you have were already answered above and in https://github.com/openssl/openssl/issues/10902.

To summarize what I think is not that hard to achieve consensus on inititally:
* the initial goal is to add a second build system that works in parallel to existing one; whether it can/will replace the original will be seeing later and will be a major breaking change to OpenSSL, it is out of scope of the initial proposal
* definitely getting rid of Perl for end-users, but there could be something in OpenSSL's CI that uses Perl in the meantime to make this viable more quickly, which is already achieved in Meson port I worked on and example in the very first post here doesn't use Perl; removing Perl completely is possible although difficult and out of scope of the initial proposal
* since this will be an additional build system, platform support is not as critical and can be expanded over time based on user feedback, Meson port already supports a substantial list of CPU architectures and operating systems and is a decent start
* there will 100% be CI integration for it, no debate about it in 2024
* having the same source for files for both build system is reasonable and possible
* community already maintains OpenSSL wrap for Meson, GYP version in Node.js's repp, probably multiple CMake integrations and countless other things, so I'm confident it will not be hard to see a few folks that can occasionally help with things specific to new build system to prevent blockers related to it while maintainers familiarize themselves with it; in fact I found Meson community very welcoming and helpful so far

I agree it is easier to review PR that is already created, but at the same time please understand that it is hard to get someone committed to creating such a non-trivial PR that so far maintainers seem to be against it to begin with. In the end many of us are not paid to do this on both sides.

If the official stance changes to accepting a second build system support such as Meson upstream assuming a specific set of conditions are met, that opens up a design space that community can work with. It just wasn't the case previously and I really appreciate above messages with specific questions/concenrs.

**Comment by Croydon** [comment]
I fully agree with everything @nazar-pc said.

The only question would then be how we decide CMake vs Meson as adding both seems excessive, people tend to have passionate opinions about their favorites and leaving it to a race as to who can create the first pull request is kinda silly.

**Comment by nhorman** [comment]
Ok, so it sounds like we have consensus on the following:

1) a new build system would exist as secondary build system, implying the following:
    - That it not break the existing build system (as that is the official build system for openssl)
    - That it use the existing build dependency inputs (to avoid needing to update multiple dependency files)
2)  That, as a secondary build system it need not cover all openssl supported primary/secondary/community platforms initially (though expanding to that would be great, and a prerequisite to any conversion to this being the official build system)
3) That the general removal of perl is a large goal beyond the scope of any potential pr, at least in the immediate future until such time as a new build system is created and merged as a secondary build system.
4) That CI for this PR is imperative, to avoid subsequent bit rot.  To be clear here, when I say CI, I'm referring specifically to Openssl CI.  As a secondary build environment, I would expect that any subsequently created CI jobs in a future PR be addative (that is to say, since this environment will not cover all of our platforms, the existing build system will be used for current jobs, and new jobs will be added simply to exercise the new build environment on platforms that it can target)
5) While it is a state of being a secondary build environment, the maintenance of this environment will be a community effort.  Openssl development will gladly help as time allows, but won't be a guaranteed priority.

That feels like pretty good consensus to me.

If you are now comfortable, shall I assign this to you, @nazar-pc or you @Croydon as a community effort?

**Comment by michael-o** [comment]
Just wanted tp add my notes as a bystander: I do use OpenSSL on HP-UX for the past 15+ years. The build system works. For our engineering code in Fortran (large codebase) we had to stick with CMake (from homegrown) because there is nothing else available and even that is a problem because CMake dropped HP-UX support in 3.10 because we wanted C+11 with libuv. Guess what aCC does not support C11/C++11. I had a very long discussion with HPE over couple of years and end of story: that support will never happen.

Upshot: So if CMake -- as a very popular build file generator -- has limited platform support Meson will support even less. I am grateful to the OpenSSL folks making this a no-brainer for me (therefore I fixed some HP-UX related stuff recently). OpenSSL thrives to cover as many systems as possible, many build system won't.

**Comment by nhorman** [comment]
@michael-o your concerns are heard, wide ranging platform support is a priority for our community and the core development team, and as such, any migration of openssl to a new build system is incumbent on it working on our full existing platform list (primary/secondary/community).

That said, reviewing that list, I do note that HP-UX is listed as an unadoped platform.  Being in that status, places openssl at a disadvantage regarding support maintenance.  It would be unreasonable to expect any future development updates to guarantee compatibility going forward on platforms that don't receive regular testing within the community.  I would strongly encourage you, if HP-UX is to be a long term target you have interest in, to identify a community platform person for those targets  (be it you or someone else in your organization), to volunteer to support the target, so that we can move it to being community supported.  Doing so would create a greater sense of consistency for your needs.

**Comment by nhorman** [comment]
based on comments from @nazar-pc in #16188 I'm marking this as a community issue, @nazar-pc please assign to yourself or an identified developer as soon as possible

**Comment by eli-schwartz** [comment]
> The only question would then be how we decide CMake vs Meson as adding both seems excessive, people tend to have passionate opinions about their favorites and leaving it to a race as to who can create the first pull request is kinda silly.

Choosing multiple systems (generating both meson.build and CMakeLists.txt) from `build.info` files, assuming the results would be included in the official tarball releases, would have one particular advantage:

Users of both meson and cmake could integrate openssl into their own build systems, for application distribution on platforms where openssl isn't natively available, or for building and shipping static binaries. Windows springs to mind as a good example of this need. It's a general topic of significant interest, for *libraries* rather than *applications* (because applications don't have to worry about the larger ecosystem they are built in, due to being "leaf package" software).

If the community interest is there from both camps, and people are willing to maintain it, I see no particular reason why both can't be supported. (I say that even though I dislike cmake on a personal level.)

**Comment by eli-schwartz** [comment]
> Upshot: So if CMake -- as a very popular build file generator -- has limited platform support Meson will support even less. I am grateful to the OpenSSL folks making this a no-brainer for me (therefore I fixed some HP-UX related stuff recently). OpenSSL thrives to cover as many systems as possible, many build system won't.

If python 3.7 can run on it (and it does) then there is no reason meson cannot do so in principle as well. There will be no CI for it unless someone volunteers machines to do so but we are happy to accept PRs and work with interested consumers.

It probably does "mostly work" with a GCC port, the biggest source of complexity is always recognizing compilers and how to interface with them -- and especially recognizing compilers that are specific to a platform i.e. vendor-provided compilers.

We've gotten patches from AIX users who are building e.g. numpy/scipy, for example, even though that's not something we have available ourselves to test with.

**Comment by thiagomacieira** [comment]
Commenting on Qt's selection of CMake:

The decision seemed non-transparent because there was effectively only one choice. There was nothing to _choose_. The only contender that even had people offering to help us support the buildsystem was qbs, but that failed one of the preconditions I had listed: another reasonably-complex project adopting it and proving the capabilities.

Since then, I have become involved with another project that uses Meson (https://github.com/opendcdiag/opendcdiag) and I have to say I am impressed with how easy it is to convert from Autoconf-style projects or to start from scratch. The problem I find is that the moment you want to do anything complex, you can't: Meson doesn't offer macros so you could do some thing repeatedly. For a moderately complex project, you have to either do a lot of copy & paste, or implement in the meson sources themselves, in Python. Therefore, I conclude Meson would not have supported Qt at all; all you have to do is look at the number of macros we've had to create for CMake to support building Qt itself and building applications against Qt.

I don't know whether the same applies to OpenSSL or not. It may suffice for your needs.

Another aspect for you to consider is compatibility. The last version that supported Python 3.7 (which is what CentOS 7 comes with) was 0.61 and that might still be a target of yours. I do point out that the standard CMake is equally ancient, but there's a 3.14 in EPEL, whereas I don't think there is a Python update there or in devtoolset.

**Comment by thiagomacieira** [comment]
> Users of both meson and cmake could integrate openssl into their own build systems, for application distribution on platforms where openssl isn't natively available, or for building and shipping static binaries. Windows springs to mind as a good example of this need.

I would advise against writing a buildsystem-generator tool. If you're going to convert to something, that something is what you should maintain.

Speaking of Windows, that's also another point in favour of CMake: Visual Studio comes with CMake, but not Meson (or Python).

**Comment by eli-schwartz** [comment]
> Another aspect for you to consider is compatibility. The last version that supported Python 3.7 (which is what CentOS 7 comes with) was 0.61 and that might still be a target of yours. I do point out that the standard CMake is equally ancient, but there's a 3.14 in EPEL, whereas I don't think there is a Python update there or in devtoolset.

Meson 0.61 was the last version that dropped support for python 3.6 -- as for 3.7, that is our current minimum.

We have no ETA on when we'll drop support for it, I personally dislike dropping support for even EOL versions of python unless doing so provides specific measurable benefits to meson development, and nothing in 3.7 has really grabbed my attention as something we need to upgrade the baseline for. When we dropped 3.6 support we gained the ability to use dataclasses and type annotations w/ future annotations which were both *incredibly* beneficial to wrangling the codebase, but there's no killer feature I'm looking forward to after that. (In 2026 we could get some nice features from python 3.10 but it is *way* too soon to think about that.)

**Comment by eli-schwartz** [comment]
> Speaking of Windows, that's also another point in favour of CMake: Visual Studio comes with CMake, but not Meson (or Python).

Note that Visual Studio doesn't need to come with python, since current versions of Windows come with python as a Store App redirector (it downloads the first time you run it, if I understand correctly) as part of the base operating system, not the Visual Studio addon.

**Comment by thiagomacieira** [comment]
Sorry, I was off by one then. CentOS 7 has Python 3.6. That's why OpenDCDiag can't upgrade from Meson 0.61.

> Note that Visual Studio doesn't need to come with python, since current versions of Windows come with python as a Store App redirector (it downloads the first time you run it, if I understand correctly) as part of the base operating system, not the Visual Studio addon.

Interesting, I didn't know that. I can confirm that it offers to install Python 3.12 if you just type "python" from the VS 2022 command prompt.

**Comment by michael-o** [comment]
> @michael-o your concerns are heard, wide ranging platform support is a priority for our community and the core development team, and as such, any migration of openssl to a new build system is incumbent on it working on our full existing platform list (primary/secondary/community).
> 
> That said, reviewing that list, I do note that HP-UX is listed as an unadoped platform. Being in that status, places openssl at a disadvantage regarding support maintenance. It would be unreasonable to expect any future development updates to guarantee compatibility going forward on platforms that don't receive regular testing within the community. I would strongly encourage you, if HP-UX is to be a long term target you have interest in, to identify a community platform person for those targets (be it you or someone else in your organization), to volunteer to support the target, so that we can move it to being community supported. Doing so would create a greater sense of consistency for your needs.

As one of the maintainers of Apache Tomcat and Tomcat Native I do test the OpenSSL C shim on many platforms, including HP-UX because I use it at work. I do compile test builds once in while. Here is the *but*: HPE delivers OpenSSL also as part of their OS for all crypto including OpenSSH. I am 99% certain that they never upstreamed their changes I would see as their obligation to do that since peole are paying $$$.
But rest assured that I , as a avid OSS contributor, will provide patches when I face issues and are able to fix them myself on that platform. I have done this for numerous projects on HP-UX solve my needs. For now, I am happy with your HP-UX support.

**Comment by vtorri** [comment]
note that  `muon` (c99 rewrite of `meson`, as mentioned by @eli-schwartz ) has a bootstrap level (`muon` can build itself) and also provides a c99 ninja-like build system called `samu` (built from [samurai).](https://github.com/michaelforney/samurai)). And `muon` is quite faster than `meson`

so you have a full c99 build system with muon.

**Comment by theoparis** [comment]
It is worth noting that both muon and samurai lack a lot of features compared to ninja and meson. For example, muon cannot build GTK, and samurai cannot build projects such as vulkan-headers.

**Comment by dcbaker** [comment]
Muon tends to grow features more when there's a compelling reason to add them. Something like "OpenSSL needs this" is the kind of reason things get added to Muon.

**Comment by eli-schwartz** [comment]
> For example, muon cannot build GTK, and samurai cannot build projects such as vulkan-headers.

vulkan-headers works for me though? In general, the thing that samurai is *lacking* is just... support for Fortran (and C++20 modules) dyndeps.

Regarding muon, it actually does successfully configure gtk if you discount use of gnome.post_install. It doesn't build, but it does configure! This is due to https://github.com/muon-build/muon/blob/master/src/script/modules/gnome.meson which is still pretty WIP. But yes, muon's support coverage is mostly notable for lacking robust module implementations, which doesn't matter to all projects. It's getting better in part by proxying that out to builtin scripts.

**Comment by vtorri** [comment]
There is also muon, which is a C99 port of meson, which includes also a ninja-like build system.

**Comment by eli-schwartz** [comment]
> I don't think the problem is as much meson's dependencies but Python normally depends on OpenSSL. There are massive amounts of standard library disabled if you don't build Python against OpenSSL. I can't recollect with 100% certainty but iirc Meson cannot work with Python that has been built without OpenSSL support. Cyclic dependencies are quite inconvenient for distros.

Meson does not explicitly use networking except for optional "wrap dependency" files, which know how to recognize a cpython build lacking OpenSSL support and raise a suitable warning indicating that downloads aren't available (including a mention of a command line flag to "allow-insecure" and use urllib with SSL disabled, keeping in mind that downloads are checksummed as well).

It does make reasonably heavy use of hashlib, but that should always work even if it's slower in the absence of openssl optimized routines...

I'm not immediately familiar with any other downsides of configuring cpython without openssl support. It is in theory quite doable using a two-stage bootstrap (build ring 0 python without SSL support). CPython is anyways in the ring 0 bootstrap because glibc depends on it, but I'm not sure whether that means distros also build a stripped down copy without SSL  (i.e. is openssl also in the ring 0 bootstrap).

**Comment by nanonyme** [comment]
> > I don't think the problem is as much meson's dependencies but Python normally depends on OpenSSL. There are massive amounts of standard library disabled if you don't build Python against OpenSSL. I can't recollect with 100% certainty but iirc Meson cannot work with Python that has been built without OpenSSL support. Cyclic dependencies are quite inconvenient for distros.
> 
> Meson does not explicitly use networking except for optional "wrap dependency" files, which know how to recognize a cpython build lacking OpenSSL support and raise a suitable warning indicating that downloads aren't available (including a mention of a command line flag to "allow-insecure" and use urllib with SSL disabled, keeping in mind that downloads are checksummed as well).
> 
> It does make reasonably heavy use of hashlib, but that should always work even if it's slower in the absence of openssl optimized routines...
> 
> I'm not immediately familiar with any other downsides of configuring cpython without openssl support. It is in theory quite doable using a two-stage bootstrap (build ring 0 python without SSL support). CPython is anyways in the ring 0 bootstrap because glibc depends on it, but I'm not sure whether that means distros also build a stripped down copy without SSL (i.e. is openssl also in the ring 0 bootstrap).

Would it be a problem that there is no hashlib?

**Comment by nanonyme** [comment]
For alternative build system Meson sounds fine though if the project wants to support multiple build systems.

**Comment by eli-schwartz** [comment]
> Would it be a problem that there is no hashlib?
"hashlib" isn't missing. The C speedup "_hashlib" is -- it will fall
back to importing:

_sha1
_md5
_sha2
_blake2
_sha3

modules, as a fallback from __get_openssl_constructor erroring out and
failing over to __get_builtin_constructor.

Anything in algorithms_guaranteed should always be unconditionally
available whether you have SSL support or not. It may be slower, but it
will be there.

We only use sha1 and sha256, except via the fs module that only
guarantees a subset of algorithms_guaranteed.

https://mesonbuild.com/Fs-module.html#hash

**Comment by BwL1289** [comment]
Looks like the wrapdb version hasn't been updated in over 3 years: https://github.com/mesonbuild/wrapdb/blob/master/subprojects/openssl.wrap.

Is there interest in supporting this upstream?

**Comment by nirbheek** [comment]
The wrapdb wrap can be updated if upstream shows interest. I know that @amyspark is interested in this work.

**Comment by tp-m** [comment]
> The wrapdb wrap can be updated if upstream shows interest. I know that [@amyspark](https://github.com/amyspark) is interested in this work.

There are recent PRs against wrapdb to update the wrap fwiw:
- https://github.com/mesonbuild/wrapdb/pull/2554
- https://github.com/mesonbuild/wrapdb/pull/2308

**Comment by SjoerdNijboer** [comment]
> If you want to remove the need for perl, you will need to either
> have a way so that it's not always needed during a build, or have
> also have a replacement/alternative implementation.

I've been experimenting with AI a bit and have generated equivalent Make + C++ code to create small applications that emit byte-equivalent .S files.
The verification step on the result was done with running both in docker files.
It's about ~250.000 LOC added for an absolute equivalent asm output for all targets.
That still leaves the question open if it should be done in the first place, and on how this should be reviewed IF ever accepted. (I'm not advocating to bypass regular pull-requests or create huge pull-requests. this was just an experiment on how far I could push a too.)
But I just want to say that nowdays this is possible and shouldn't be a blocker for moving from one build system to another.

At that point I think it becomes a legitimate question if the future of any openssl buildsystem should contain any perl at all.

**Comment by nanonyme** [comment]
250 000 LOC AI generated ASM code sounds like a terrible idea for a security-critical project.

**Comment by SjoerdNijboer** [comment]
Absolutely,

But it's more of an FYI that it's possible.
That, and I personally hold the belief that generated code should just go through thesame steps that regular code that gets checked in should go through.
My point is that it's absolutely possible to remove perl from the codebase for something else with the tools of today IF people want that.
It doesn't have to be a multi-year effort for a team anymore.
It also shouldn't be a 3-day thing where someone checks something, presses merge and then all code becomes garbage.
I'm just convinced and wanted to communicate that this is one of the very few tasks where code generation actually can be done since:
1) no runtime code gets altered
2) there's a known version with known behaviour already (the existing buildsystem)
3) it's possible to write some automated tests that the perl-asm for all platforms should be byte-equal between perl generated output and any other system generated output.

Also please note that I haven't linked any branch or forked OpenSSL at all.
I'm also not asking to merge it.
I'm not concidering this generated code worth showing because we all have no trust in it.
It's just... performing this task might be a bit smaller than initially thought

**Comment by nanonyme** [comment]
My point of doing full security review on 250 000 of ASM sounds like a nightmare. Based on history writing code is not the bottleneck in most OSS projects, reviewing it is. With crypto like OpenSSL it's even worse. Minor bugs might be impossible to spot from that magnltude of submission and in crypto minor bugs can lead into major security vulnerabilities. Unless idea is to blindly accept contribution based on passing tests without anyone in project reading and understanding the relevant code which I doubt is an acceptable outcome.

**Comment by SjoerdNijboer** [comment]
> Based on history writing code is not the bottleneck in most OSS projects, reviewing it is.

That's fair, I hadn't considered that.
I figured with specifically this ticket it would be different.

> Unless idea is to blindly accept contribution based on passing tests without anyone in project reading and understanding the relevant code which I doubt is an acceptable outcome.

That's never been my idea, I've seen that happen with others and I know how destructive that is for a project. Especially for what you said, is a crypto project.

Okay, then what I said is maybe fun as an experiment but absolutely irrelevant for this ticket.

**Comment by nazar-pc** [comment]
Adding 250k lines of LLM-produced assembly is no-go. The initial idea was to commit Perl-generated assembly or publish it as a release artifact, such that it doesn't need to be a build-time requirement anymore, while the project can continue to be developed as is. There were also attempts to replace Perl with Python, but leave the rest of it as is as well.

From my side though I was simply trying to avoid OpenSSL if I can at all. It is too painful to use in Rust projects, and, thankfully, it was possible to avoid in most cases.

**Comment by SjoerdNijboer** [comment]
mind you, 250k lines of LLM-produced assembly is not what I was proposing.
Right now the perl scripts generate assembly for some 6 or 7 targets on a whole bunch of compiler versions.
I was simply making the statement that replacing the per-based generator scripts is possible with say... language XYZ-based generator scripts and having it produce the exact same output with some ease.
I didn't fully appreciate the security implications of just replacing that part of the buildsystem.
I just wanted to suggest that there are a little more options today than there were when this ticket was originally created.

I have some experience re-doing buildsystems so I figured I could take a look at some projects that wanted to change buildsystems.
That's how I ended at this ticket for OpenSSL.

**Comment by nanonyme** [comment]
> Adding 250k lines of LLM-produced assembly is no-go. The initial idea was to commit Perl-generated assembly or publish it as a release artifact, such that it doesn't need to be a build-time requirement anymore, while the project can continue to be developed as is. There were also attempts to replace Perl with Python, but leave the rest of it as is as well.
> 
> From my side though I was simply trying to avoid OpenSSL if I can at all. It is too painful to use in Rust projects, and, thankfully, it was possible to avoid in most cases.

Right. I asked LLM to summarize this so I think I'm now more or less on same page. It sounds indeed Perl is *extremely* hard to replace, perlasm is just a good match for OpenSSL philosophy of having per arch optimized ASM. For the record funny you mentioned Rust as RustCrypto apparently has polar opposite approach to prefer intrinsics and avoid ASM which does lead to lower performance but is easier to reason with without complex code-generation.

**Comment by SjoerdNijboer** [comment]
> It sounds indeed Perl is _extremely_ hard to replace, perlasm is just a good match for OpenSSL philosophy of having per arch optimized ASM.

This is why I found it so interesting to give this a try. Ideally no new dependencies on additional languages should be introduced, only re-use of existing dependencies.

In my limited experience porting buildsystems, the annoying part is the custom scripted behaviour. It’ll take up most of the effort regardless of how this issue turns out. So regardless of which buildsystem is chosen, i think dealing with these scripts or replacing them should have equal quality support with a new buildsystem as it has now. And unless the goal is to introduce additional complexity I suggest they should just be replaced by a feature of a buildsystem that can natively support such generation. Specifically, I would expect any replacement buildsystem to have first-party support for compiling code, running that code to generate an output file and then using that output file to compile the main openssl project.
I don’t think Meson can do that comfortably, but I could be wrong on this.

Personally I would find it unacceptable to then also have checked-in pre-generated assembly from the current buildscripts since then adding a new target would require a lot of manual assembly tuning again. That would be loss of functionality to me.

I could for some time help figuring out what’s preferred and do some experimentation. I currently have a temporary medical situation, but my mind is completely functional. So I’m sick at home and bored.

**Comment by eli-schwartz** [comment]
> But I just want to say that nowdays this is possible and shouldn't be a blocker for moving from one build system to another.

I'm nearly entirely mostly certain that both Perl and C are turing complete programming languages and it is possible to transcribe a program written in one into the other, if one cares to do so.

The technical feasibility of doing so is relatively likely, therefore, to not have been the single biggest blocker.



> I just wanted to suggest that there are a little more options today than there were when this ticket was originally created.

I am not aware of any novel options that exist today that didn't exist in 2021.

The use of LLMs is not a novel new option. You cannot do anything with an LLM that you cannot also do without an LLM. In the most optimistic scenario, you're trading "I don't have to devote time to it" for "it will be inaccurate and bug-ridden and also cost a lot of money". And as noted above that applies to both the development and the review stages, independently. And for crypto code it is... unwise... to take a gamble on the review stage being done by something inaccurate and bug-ridden. (Actually it is unwise to take a gamble on either stage. Until the LLM plague descended on the world, it was an accepted part of the social contract that contributors do self-review, greatly easing the burden on the review stage by filtering out a lot of badness and improving the quality of the original reviewable.)



> And unless the goal is to introduce additional complexity I suggest they should just be replaced by a feature of a buildsystem that can natively support such generation. Specifically, I would expect any replacement buildsystem to have first-party support for compiling code, running that code to generate an output file and then using that output file to compile the main openssl project.
> I don’t think Meson can do that comfortably, but I could be wrong on this.

... My friend, you're taking about the build system functionality to define custom targets that run project defined commands instead of abstract built-ins like "construct an abstract executable depending on the following source files". And then using output files as inputs to another target.

It is the basic bread and butter of ANY build system.

The most complicated thing here is needing to compile executables for both build and host in a cross-compiled build, because "running that code to generate an output file" cannot execute a host machine executable if the host machine is a different operating system or CPU architecture (okay, it essentially works when crossing amd64 to i686 since the CPU is capable of executing both and operating systems are built with 32-bit emulation).

That's even the part where Meson has an advantage, since unlike certain other buildsystems I could name, Meson has an actual concept for build and host compilers available simultaneously in a single configuration so that you can run code generators at compile time.

But it's also a moot point if people don't want to move off of Perl, since perl scripts aren't architecture or CPU specific. That is in fact a real benefit of the current setup -- the OpenSSL project trades off the ability to avoid some of the fiddliest parts of cross compilation, for the downside of having a build dependency on Perl already being installed. And their configure script is already Perl, so it's *not a downside*.

It is certainly possible to move to another build system without dropping Perl as a build requirement. Just like it's possible to produce a dist tarball containing pregenerated asm (produced by perl scripts at release time) while still officially using Perl ./Configure.

**Comment by nanonyme** [comment]
> I'm nearly entirely mostly certain that both Perl and C are turing complete programming languages and it is possible to transcribe a program written in one into the other, if one cares to do so.

Technically yes though apparently in this case it seems it's not as much the language itself but support library ecosystem. 

> ... My friend, you're taking about the build system functionality to define custom targets that run project defined commands instead of abstract built-ins like "construct an abstract executable depending on the following source files". And then using output files as inputs to another target.

Fully agreed here. I don't think it makes sense to conflate Meson support and dropping Perl completely into one project.
